kubernetes-extension-test/src/test/java/org/ballerinax/kubernetes/test/IstioGatewayTest.java

```diff
@@ -0,0 +1,487 @@
+
+import static org.ballerinax.kubernetes.KubernetesConstants.DOCKER;
+import static org.ballerinax.kubernetes.KubernetesConstants.KUBERNETES;
+import static org.ballerinax.kubernetes.test.utils.KubernetesTestUtils.getDockerImage;
+
+/**
+ * Test cases for generating istio gateway artifacts.
+ *
+ * @since 0.983.0
+ */
+public class IstioGatewayTest {
+    
+    private final String balDirectory = Paths.get("src").resolve("test").resolve("resources").resolve("istio")
+            .resolve("gateway").toAbsolutePath().toString();
+    private final String targetPath = Paths.get(balDirectory).resolve(KUBERNETES).toString();
+    private final String dockerImage = "pizza-shop:latest";
+    
+    /**
+     * Build bal file with istio gateway annotations.
+     *
+     * @throws IOException               Error when loading the generated yaml.
+     * @throws InterruptedException      Error when compiling the ballerina file.
+     * @throws KubernetesPluginException Error when deleting the generated artifacts folder.
+     */
+    @Test
+    public void allFieldsTest() throws IOException, InterruptedException, KubernetesPluginException {
+        Assert.assertEquals(KubernetesTestUtils.compileBallerinaFile(balDirectory, "all_fields.bal"), 0);
+
+        // Check if docker image exists and correct
+        validateDockerfile();
+        validateDockerImage();
+
+        // Validate gateway yaml
+        File gatewayFile = Paths.get(targetPath).resolve("all_fields_istio_gateway.yaml").toFile();
+        Assert.assertTrue(gatewayFile.exists());
+        Yaml yamlProcessor = new Yaml();
+        Map<String, Object> gateway = (Map<String, Object>) yamlProcessor.load(FileUtils.readFileAsString(gatewayFile));
+        Assert.assertEquals(gateway.get("apiVersion"), "networking.istio.io/v1alpha3", "Invalid apiVersion");
+        Assert.assertEquals(gateway.get("kind"), "Gateway", "Invalid kind.");
+
+        Map<String, Object> metadata = (Map<String, Object>) gateway.get("metadata");
+        Assert.assertEquals(metadata.get("name"), "my-gateway", "Invalid gateway name");
+        Assert.assertEquals(metadata.get("namespace"), "ballerina", "Invalid gateway name");
+
+        Map<String, String> labels = (Map<String, String>) metadata.get("labels");
+        Assert.assertEquals(labels.get("label1"), "label1", "Invalid label");
+        Assert.assertEquals(labels.get("label2"), "label2", "Invalid label");
+
+        Map<String, String> annotations = (Map<String, String>) metadata.get("annotations");
+        Assert.assertEquals(annotations.get("anno1"), "anno1Val", "Invalid annotation value");
+        Assert.assertEquals(annotations.get("anno2"), "anno2Val", "Invalid annotation value");
+
+        Map<String, Object> spec = (Map<String, Object>) gateway.get("spec");
+        Map<String, Object> selector = (Map<String, Object>) spec.get("selector");
+        Assert.assertEquals(selector.get("app"), "my-gateway-controller", "Invalid selector.");
```

Comment:
Shall we use `KubernetesConstants.KUBERNETES_SELECTOR_KEY` here?

Reply:
Fixed with https://github.com/ballerinax/kubernetes/pull/220/commits/5f8887c1388816ebe18d4066c627a7d43a88bea9